File: Ontology/DataUnionOf.cs
using System.Collections.Generic;
using System.Linq;

namespace Ontology
{
    public class DataUnionOf: IDataUnionOf
    {
        private IList<IDataRange> _dataRanges;

        public DataUnionOf(
            params IDataRange[] dataRanges
            )
        {
            _dataRanges = dataRanges;
        }

        IList<IDataRange> IDataUnionOf.DataRanges => _dataRanges;

        bool IDataRange.HasMember(
            object value
            ) => _dataRanges.Any(dataRange => dataRange.HasMember(value));
    }
}

File: Web/ClientApp/src/app/Ontology/DataUnionOf.ts
import { DataRange } from "./DataRange";
import { IDataRange } from "./IDataRange";
import { IDataRangeSelector } from "./IDataRangeSelector";
import { IDataUnionOf } from "./IDataUnionOf";

export class DataUnionOf
    extends DataRange
    implements IDataUnionOf
{
    constructor(
        public DataRanges: IDataRange[]
        )
    {
        super();
    }

    HasMember(
        value: object
        ): boolean
    {
        return this.DataRanges.some(dataRange => dataRange.HasMember(value));
    }

    Select<TResult>(
        selector: IDataRangeSelector<TResult>
        ): TResult
    {
        return selector.DataUnionOf(this);
    }
}
